typing/ctype.ml

```diff
@@ -2683,24 +2684,33 @@ let complete_type_list ?(allow_absent=false) env fl1 lv2 pack2 =
                 if allow_absent then
                   complete nl fl2
                 else
-                  raise Exit
+                  raise (Missing n)
             end
         | (_, {type_arity = 0; type_kind = Type_abstract _;
                type_private = Public; type_manifest = None})
           when allow_absent ->
             complete nl fl2
-        | _ -> raise Exit
-        | exception Not_found when allow_absent->
+        | _ -> raise (Missing n)
+        | exception Not_found ->
+           if allow_absent then
             complete nl fl2
+           else raise (Missing n)
   in
   match complete fl1 pack2.pack_cstrs with
-  | res -> res
-  | exception Exit -> raise Not_found
+  | res -> Ok res
+  | exception (Missing n) -> Error n
 
 (* raise Not_found rather than Unify if the module types are incompatible *)
 let compare_package env unify_list lv1 pack1 lv2 pack2 =
   let ntl2 = complete_type_list env pack1.pack_cstrs lv2 pack2
   and ntl1 = complete_type_list env pack2.pack_cstrs lv1 pack1 in
+  let missing pos name =
+     raise_for Unify (First_class_module (Missing_type_constraint (pos,name)))
+  in
+  match ntl1, ntl2 with
+  | Error g, Ok _ -> missing First g
+  | _, Error e -> missing Second e
+  | Ok ntl1, Ok ntl2 ->
```

Comment:
Another design would be to pass the position (First | Second) to the `complete_type_list` function above, and have it call `raise_for`directly instead of an error type that is interpreted here. If you do later decide to have several different errors, this might make things simpler.